### PR TITLE
[release_tool] QA-313: New option --list-format for advanced list

### PR DESCRIPTION
**Pull request by lluiscampos**
The option supports "simple" (only repos, backwards compatible), "table"
and "json" values.

### CODE REVIEW

**Comment by lluiscampos** [comment]
Some sample calls
```
$ release_tool --list --in-integration-version 2.7.0 --list-format table
|                      COMPONENT |         VERSION |
|                      auditlogs |           1.1.0 |
|         create-artifact-worker |           1.0.2 |
|                    deployments |           2.3.0 |
|         deployments-enterprise |           2.3.0 |
|                     deviceauth |           2.6.0 |
|                   deviceconfig |           1.0.0 |
|                  deviceconnect |           1.1.0 |
|                            gui |           2.7.0 |
|                    integration |           2.7.0 |
|                      inventory |           2.3.0 |
|           inventory-enterprise |           2.3.0 |
|                         mender |           2.6.0 |
|                mender-artifact |           3.5.1 |
|                     mender-cli |           1.7.0 |
|                 mender-connect |           1.1.0 |
|                mtls-ambassador |           1.0.0 |
|                      tenantadm |           3.1.0 |
|                        useradm |          1.14.0 |
|             useradm-enterprise |          1.14.0 |
|                      workflows |           1.4.0 |
|           workflows-enterprise |           1.4.0 |
$ release_tool --list --in-integration-version 2.7.x --list-format json | jq
{
  "auditlogs": "1.1.x",
  "create-artifact-worker": "1.0.x",
  "deployments": "2.3.x",
  "deployments-enterprise": "2.3.x",
  "deviceauth": "2.6.x",
  "deviceconfig": "1.0.x",
  "deviceconnect": "1.1.x",
  "gui": "2.7.x",
  "integration": "2.7.x",
  "inventory": "2.3.x",
  "inventory-enterprise": "2.3.x",
  "mender": "2.6.x",
  "mender-artifact": "3.5.x",
  "mender-cli": "1.7.x",
  "mender-connect": "1.1.x",
  "mtls-ambassador": "1.0.x",
  "tenantadm": "3.1.x",
  "useradm": "1.14.x",
  "useradm-enterprise": "1.14.x",
  "workflows": "1.4.x",
  "workflows-enterprise": "1.4.x"
}
$ release_tool --list --in-integration-version feature-update_control_maps --list-format table
|                      COMPONENT |         VERSION |
|                      auditlogs |          master |
|         create-artifact-worker |          master |
|                    deployments | feature-update_control_maps |
|         deployments-enterprise | feature-update_control_maps |
|                     deviceauth |          master |
|                   deviceconfig |          master |
|                  deviceconnect |          master |
|                            gui |          master |
|                    integration | feature-update_control_maps |
|                      inventory |          master |
|           inventory-enterprise |          master |
|                         mender | feature-update_control_maps |
|                mender-artifact |          master |
|                     mender-cli |          master |
|                 mender-connect |          master |
|                mtls-ambassador |          master |
|                      tenantadm |          master |
|                        useradm |          master |
|             useradm-enterprise |          master |
|                      workflows |          master |
|           workflows-enterprise | feature-update_control_maps |
$ release_tool --list container --in-integration-version master --list-format table
|                      COMPONENT |         VERSION |
|               mender-auditlogs |   mender-master |
|                  mender-client |          master |
|  mender-create-artifact-worker |   mender-master |
|             mender-deployments |   mender-master |
|             mender-device-auth |   mender-master |
|            mender-deviceconfig |   mender-master |
|           mender-deviceconnect |   mender-master |
|                     mender-gui |   mender-master |
|               mender-inventory |   mender-master |
|               mender-tenantadm |   mender-master |
|                 mender-useradm |   mender-master |
|        mender-workflows-server |   mender-master |
|        mender-workflows-worker |   mender-master |
|                mtls-ambassador |   mender-master |
$ release_tool --list docker --in-integration-version master --list-format table
|                      COMPONENT |         VERSION |
|                      auditlogs |   mender-master |
|         create-artifact-worker |   mender-master |
|                    deployments |   mender-master |
|         deployments-enterprise |   mender-master |
|                     deviceauth |   mender-master |
|                   deviceconfig |   mender-master |
|                  deviceconnect |   mender-master |
|                            gui |   mender-master |
|                      inventory |   mender-master |
|           inventory-enterprise |   mender-master |
|           mender-client-docker |          master |
|    mender-client-docker-addons |          master |
|             mender-client-qemu |          master |
|        mender-client-qemu-rofs |          master |
|                mtls-ambassador |   mender-master |
|                      tenantadm |   mender-master |
|                        useradm |   mender-master |
|             useradm-enterprise |   mender-master |
|                      workflows |   mender-master |
|           workflows-enterprise |   mender-master |
|    workflows-enterprise-worker |   mender-master |
|               workflows-worker |   mender-master |
```

**Comment by lluiscampos** [comment]
One "known" bug, though: `list` always inspects local checkout so a query with an old version will produce something like this:
```
$ release_tool --list container --in-integration-version 2.6.1 --list-format table
|                      COMPONENT |         VERSION |
|               mender-auditlogs |    mender-2.6.1 |
|                  mender-client |           2.5.1 |
|  mender-create-artifact-worker |    mender-2.6.1 |
|             mender-deployments |    mender-2.6.1 |
|             mender-device-auth |    mender-2.6.1 |
|            mender-deviceconfig |                 |
|           mender-deviceconnect |    mender-2.6.1 |
|                     mender-gui |    mender-2.6.1 |
|               mender-inventory |    mender-2.6.1 |
|               mender-tenantadm |    mender-2.6.1 |
|                 mender-useradm |    mender-2.6.1 |
|        mender-workflows-server |    mender-2.6.1 |
|        mender-workflows-worker |    mender-2.6.1 |
|                mtls-ambassador |    mender-2.6.1 |
```
Which basically says that deviceconfig is not released in 2.6.1 (instead of just removing it from the list). This is pre-existing, so we can fix it in a different PR, if any.

**Review by kacf — APPROVED** [review]
Nice!

There must be a way we can monetize release_tool, it's getting to be so nice! :moneybag: :smile:

File: extra/release_tool.py [hunk 1]
```diff
@@ -656,17 +656,42 @@ def do_list_repos(args, optional_too, only_backend):
...
+    )
+
+    repos = Component.get_components_of_type(
+        type,
+        only_release=(not optional_too),
+        only_non_independent_component=(only_backend),
+    )
+    repos.sort(key=lambda x: x.name)
+
+    row_format = "| {:>30} | {:>15} |"
```
Comment (oleorhagen) [inline extra/release_tool.py:670]: Suggestion:
It'd be nice if the second column was formatted by the length of the longest entry

**Review by oleorhagen — APPROVED** [review]

File: extra/release_tool.py (hunk 1 above)
Comment (lluiscampos) [reply extra/release_tool.py:670]: Roger that :+1:

**Review by mzedel — APPROVED** [review]
👍 thanks for that! I'm already looking forward to consuming the resulting json 🎉

**Comment by lluiscampos** [comment]
@oleorhagen Reworked it a bit to get calculate the max len of each column.

Related to it, I actually moved the logic into two steps: first get the versions then print stuff. I think it actually resulted in much cleaner code, so thanks for that.

New formatting:
```
$ release_tool --list --in-integration-version 2.7.0 --list-format table
|              COMPONENT | VERSION |
|              auditlogs |  1.1.0 |
| create-artifact-worker |  1.0.2 |
|            deployments |  2.3.0 |
| deployments-enterprise |  2.3.0 |
|             deviceauth |  2.6.0 |
|           deviceconfig |  1.0.0 |
|          deviceconnect |  1.1.0 |
|                    gui |  2.7.0 |
|            integration |  2.7.0 |
|              inventory |  2.3.0 |
|   inventory-enterprise |  2.3.0 |
|                 mender |  2.6.0 |
|        mender-artifact |  3.5.1 |
|             mender-cli |  1.7.0 |
|         mender-connect |  1.1.0 |
|        mtls-ambassador |  1.0.0 |
|              tenantadm |  3.1.0 |
|                useradm | 1.14.0 |
|     useradm-enterprise | 1.14.0 |
|              workflows |  1.4.0 |
|   workflows-enterprise |  1.4.0 |
$ release_tool --list container --in-integration-version master --list-format table
|                     COMPONENT |       VERSION |
|              mender-auditlogs | mender-master |
|                 mender-client |        master |
| mender-create-artifact-worker | mender-master |
|            mender-deployments | mender-master |
|            mender-device-auth | mender-master |
|           mender-deviceconfig | mender-master |
|          mender-deviceconnect | mender-master |
|                    mender-gui | mender-master |
|              mender-inventory | mender-master |
|              mender-tenantadm | mender-master |
|                mender-useradm | mender-master |
|       mender-workflows-server | mender-master |
|       mender-workflows-worker | mender-master |
|               mtls-ambassador | mender-master |
$ release_tool --list --in-integration-version feature-update_control_maps --list-format table
|              COMPONENT |                     VERSION |
|              auditlogs |                      master |
| create-artifact-worker |                      master |
|            deployments | feature-update_control_maps |
| deployments-enterprise | feature-update_control_maps |
|             deviceauth |                      master |
|           deviceconfig |                      master |
|          deviceconnect |                      master |
|                    gui |                      master |
|            integration | feature-update_control_maps |
|              inventory |                      master |
|   inventory-enterprise |                      master |
|                 mender | feature-update_control_maps |
|        mender-artifact |                      master |
|             mender-cli |                      master |
|         mender-connect |                      master |
|        mtls-ambassador |                      master |
|              tenantadm |                      master |
|                useradm |                      master |
|     useradm-enterprise |                      master |
|              workflows |                      master |
|   workflows-enterprise | feature-update_control_maps |
```

**Comment by lluiscampos** [comment]
And now left-aligned with lots of love for @oleorhagen

**Comment by oleorhagen** [comment]
https://media.giphy.com/media/10cwmeKKLtrkS4/giphy.gif

**Pull request merged**